toolchain/sem_ir/bind_name.h

```diff
@@ -0,0 +1,74 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_BIND_NAME_H_
+#define CARBON_TOOLCHAIN_SEM_IR_BIND_NAME_H_
+
+#include "common/hashing.h"
+#include "toolchain/base/value_store.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::SemIR {
+
+struct BindNameInfo : public Printable<BindNameInfo> {
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "{name: " << name_id << ", enclosing_scope: " << enclosing_scope_id
+        << ", index: " << bind_index << "}";
+  }
+
+  // The name.
+  NameId name_id;
+  // The enclosing scope.
+  NameScopeId enclosing_scope_id;
+  // The index for a compile-time binding. Invalid for a runtime binding.
+  CompileTimeBindIndex bind_index;
+};
+
+// Hashing for BindNameInfo.
+inline auto CarbonHashValue(const BindNameInfo& value, uint64_t seed)
+    -> HashCode {
+  Hasher hasher(seed);
+  hasher.Hash(value);
+  return static_cast<HashCode>(hasher);
+}
+
+// DenseMapInfo for BindNameInfo.
+struct BindNameInfoDenseMapInfo {
+  static auto getEmptyKey() -> BindNameInfo {
+    return BindNameInfo{.name_id = NameId(-1),
```

Comment:
Changed to be different from `NameId::Invalid` to avoid colliding with a possible value.

Reply:
Just to note -- I'm not seeing how you're doing this, but would it be worth adding NameId::HashAsEmpty, HashAsTombstone so that it's clearer on the NameId that these need to be reserved values?

Reply:
I ended up switching to storing the intentionally-bad value in the compile-time binding index, where I think the risk of having more reserved values is much lower. I understand the motivation behind adding names for the extra states, but that would seem to legitimize using those values as valid values, and the point here is to form a representation that's not a valid value, so I'm hesitant. I've added a comment to the declaration of the type to further reduce the risk.